File: back/app/schemas/book.py
from typing import Optional, List
from pydantic import BaseModel
from .category import Category
# =========================
# Base
# =========================

class BookBase(BaseModel):
    title: str
    author: Optional[str] = None
    description: Optional[str] = None
    isbn: Optional[str] = None
    published_year: Optional[int] = None
    category_id: Optional[int] = None
    quantity: Optional[int] = 1
    cover_url: Optional[str] = None

class RecommendedBook(BaseModel):
    id: int
    title: str
    author: Optional[str] = None
    description: Optional[str] = None
    isbn: Optional[str] = None
    published_year: Optional[int] = None
    category_id: Optional[int] = None
    quantity: Optional[int] = 1
    cover_url: Optional[str] = None

    model_config = {
        "from_attributes": True
    }

# =========================
# Création
# =========================
class BookCreate(BookBase):
    tags: Optional[List[int]] = []   # IDs des tags


# =========================
# Mise à jour
# =========================
class BookUpdate(BaseModel):
    title: Optional[str] = None
    author: Optional[str] = None
    description: Optional[str] = None
    isbn: Optional[str] = None
    published_year: Optional[int] = None
    category_id: Optional[int] = None
    quantity: Optional[int] = None
    cover_url: Optional[str] = None
    tags: Optional[List[int]] = None

# =========================
# Réponse (lecture)
# =========================
class BookResponse(BookBase):
    id: int
    
    model_config = {
        "from_attributes": True
    }

class TagSimple(BaseModel):
    id: int
    name: str

    model_config = {
        "from_attributes": True
    }


class BookResponseWithRelations(BookResponse):
    category: Optional[Category]
    tags: List[TagSimple] = []
    # loans: List[LoanResponse] = []


class BookFilter(BaseModel):
    title: Optional[str] = None
    author: Optional[str] = None
    published_after: Optional[int] = None
    published_before: Optional[int] = None
    category_ids: Optional[List[int]] = None
    tag_ids: Optional[List[int]] = None

File: front/src/app/interfaces/book.ts
// book.model.ts

import { Category } from "./category";
import { Tag } from "./tag";

export interface BookCategoryLink {
  category_id: number;
}

export interface Book {
  id: number;
  title: string;
  author?: string;
  description?: string;
  isbn?: string;
  published_year?: string;
  quantity: number;
  cover_url?: string;

  category_id?: number;
  tag?: Tag[];
}

export interface BookFilter {
  title?: string;
  author?: string;
  published_after?: number;
  published_before?: number;
  category_ids?: number[];
  tag_ids?: number[];
}

export interface BorrowedBook {
  loan_id: number;
  book_id: number;
  title: string;
  author: string;
  description: string;
  isbn: string;
  published_year: number;
  category_id: number;
  cover_url: string;
  loan_date: string;
  due_date: string;
  return_date?: string;
  status: string;
  book_quantity?: number;
}
